### PR TITLE
Scala sealed trait ability (mostly) in dataconf using dataclass

### PR DESCRIPTION
**Pull request by dwsmith1983**
@zifeo added the ability for nested dataclass methods (sealed trait pureconfig functionality in Scala), updated README, incremented to 0.1.6, and created version.py which will provide __version__ and get the latest from the pyproject.toml

Also, on my local, the following test always fails. How do I get it to pass?
```python
tests/test_parser.py:186 (TestParser.test_missing_type)
self = <tests.test_parser.TestParser object at 0x7fe398c417c0>

    def test_missing_type(self) -> None:
    
        with pytest.raises(MissingTypeException):
>           loads("", Dict)
E           Failed: DID NOT RAISE <class 'dataconf.exceptions.MissingTypeException'>
```

### CODE REVIEW

**Comment by dwsmith1983** [comment]
@zifeo the lint test failed. Not sure why
```
Cache Size: ~13 MB (13738048 B)
/usr/bin/tar --use-compress-program zstd -d -xf /home/runner/work/_temp/a9c1d847-2022-40d2-a351-a81aa3882ad2/cache.tzst -P -C /home/runner/work/dataconf/dataconf
Error: The operation was canceled.
```

**Comment by dwsmith1983** [comment]
@zifeo I see that from `dataconf.version import __version__` is the failing behavior for `zimports`. Is there anyway to add this to a safe list? This one allows anyone to call the version (pretty common)
```python
import dataconf
print(dataconf.__version__)
```

**Comment by zifeo** [comment]
@dwsmith1983 What kind of failure do you get from the subclass approach? Currently if you use the union type, it does not use https://github.com/zifeo/dataconf/pull/10/files#diff-4198c13639f9dd31ef7b259b7e7167468728b79898427be0fa2e3be56f4526acR116-R127, right?

**Comment by dwsmith1983** [comment]
> @dwsmith1983 What kind of failure do you get from the subclass approach? Currently if you use the union type, it does not use https://github.com/zifeo/dataconf/pull/10/files#diff-4198c13639f9dd31ef7b259b7e7167468728b79898427be0fa2e3be56f4526acR116-R127, right?

With union, correct it skips it. With `__subclass__()`, `clazz == []`. It looks like the only the usage of `InputType()` so far allows for single type parameter since `dir(clazz)` will bring up `InputType`. Any ideas? In some cases, `InputType()` could have many subclasses so automatic parsing would be nicer for the user then a long list.

**Comment by zifeo** [comment]
@dwsmith1983 I was thinking of something like
```python
class Parent:
  pass

@dataclass
class StringImpl(Parent):
   name: Text
   age: Text

@dataclass
class IntImpl(Parent):
  area_code: int
  phone_num: Text

@dataclass
 class Base:
     location: Text
     input_source: Parent

str_conf = """
 {
     location: Europe
     input_source {
         name: Thailand
         age: "12"
     }
 }
 """
```

**Comment by dwsmith1983** [comment]
@zifeo wouldn't need to make parent then an ABC class which would be nice with the abstract methods and overriding. That would be seamless transition from Scala/Java hocon with traits to Python.

https://docs.python.org/3/library/abc.html

**Comment by zifeo** [comment]
@dwsmith1983 not really fan of the `register` syntax, but it could support definitely both!

**Comment by dwsmith1983** [comment]
@zifeo have you tested the syntax you suggested? I just gave it try. I didn't step through it but from the error, it looks like the same is occurring `dir(clazz) == []`. Did it work for you?

**Comment by zifeo** [comment]
@dwsmith1983 can you push your changes? I will give it a try.

**Comment by dwsmith1983** [comment]
@zifeo pushed

**Comment by dwsmith1983** [comment]
@zifeo any updates?

**Comment by zifeo** [comment]
What about the following?
```
diff --git a/dataconf/utils.py b/dataconf/utils.py
index cb1082a..4fd3ee3 100644
--- a/dataconf/utils.py
+++ b/dataconf/utils.py
@@ -118,14 +118,19 @@ def __parse(value: any, clazz, path):
     #       if subclasses are a dataclass parse values
     #       the idea here is to replicate parsing of a sealed trait in Scala
     #       when using pureconfig
-    for key in dir(clazz):
-        nested_class = getattr(clazz, key)
-        if is_dataclass(nested_class):
-            field_keys = set(
-                [i.name for i in fields(getattr(clazz, nested_class.__name__))]
-            )
-            if field_keys == set(value.keys()):
-                return __parse(value, getattr(clazz, nested_class.__name__), "")
+    child_failures = []
+    for child_clazz in sorted(clazz.__subclasses__(), key=lambda c: c.__name__):
+        if is_dataclass(child_clazz):
+            try:
+                return __parse(value, child_clazz, path)
+            except TypeConfigException as f:
+                child_failures.append(f)
+
+    if len(child_failures) > 0:
+        fails = '\n- '.join(child_failures)
+        raise TypeConfigException(
+            f"expected type {clazz} at {path}, failed subclasses:{fails}"
+        )
 
     raise TypeConfigException(f"expected type {clazz} at {path}, got {type(value)}")
 
diff --git a/tests/test_parser.py b/tests/test_parser.py
index 32b2088..87d280a 100644
--- a/tests/test_parser.py
+++ b/tests/test_parser.py
@@ -249,7 +249,7 @@ class TestParser:
         @dataclass
         class Base:
             location: Text
-            input_source: InputType()
+            input_source: InputType
 
         str_conf = """
         {
```

**Comment by dwsmith1983** [comment]
@zifeo can you push the commit so I can pull it and try it out?

File: dataconf/utils.py [hunk 1]
```diff
@@ -113,6 +113,26 @@ def __parse(value: any, clazz, path):
...
+    child_failures = []
+    for child_clazz in sorted(clazz.__subclasses__(), key=lambda c: c.__name__):
+        if is_dataclass(child_clazz):
+            try:
+                return __parse(value, child_clazz, path)
+            except TypeConfigException as f:
+                child_failures.append(str(f))
+
+    # no need to check length; false if empty
+    if child_failures:
```
Comment (dwsmith1983) [inline dataconf/utils.py:130]: Here I removed the length check

File: dataconf/utils.py [hunk 2]
```diff
@@ -113,6 +113,26 @@ def __parse(value: any, clazz, path):
...
+    #       if subclasses are a dataclass parse values
+    #       the idea here is to replicate parsing of a sealed trait in Scala
+    #       when using pureconfig
+    child_failures = []
+    for child_clazz in sorted(clazz.__subclasses__(), key=lambda c: c.__name__):
+        if is_dataclass(child_clazz):
+            try:
+                return __parse(value, child_clazz, path)
+            except TypeConfigException as f:
+                child_failures.append(str(f))
```
Comment (dwsmith1983) [inline dataconf/utils.py:127]: @zifeo also changed `f` to `str(f)` for the join. We will need a test for failure too to make sure all is working.

**Review by zifeo — APPROVED** [review]
Let's update the doc and all good.

**Comment by dwsmith1983** [comment]
@zifeo updated doc and added a fail test check for the abstract class trait

**Pull request merged**